File: pkg/engine/engine.go
package engine

import "warframe-state-graph/pkg/model"

// ResolveState はノードの充足状態を再帰的に判定する。
// requires が空、または全て SATISFIED なら ACTIONABLE。循環参照は BLOCKED 扱いで打ち切る。
func ResolveState(g *model.Graph, nodeID string, stack map[string]bool) model.NodeState {
	node, ok := g.Nodes[nodeID]
	if !ok {
		return model.StateBlocked
	}
	if node.Satisfied {
		return model.StateSatisfied
	}
	if stack[nodeID] {
		return model.StateBlocked // 循環防御
	}
	if len(node.Requires) == 0 {
		return model.StateActionable
	}

	next := make(map[string]bool, len(stack)+1)
	for k := range stack {
		next[k] = true
	}
	next[nodeID] = true

	for _, reqID := range node.Requires {
		if ResolveState(g, reqID, next) != model.StateSatisfied {
			return model.StateBlocked
		}
	}
	return model.StateActionable
}

// CascadeSatisfyRequires は、あるノードが達成済みになった時、その requires チェーンを
// 再帰的に遡って全て達成済みにする。「後の工程が終わっているなら前提工程も終わっているはず」
// という requires（前提条件）特有の意味論に基づく（例: Natah を達成したなら The War Within や
// Saya's Vigil 等の前提クエストは必然的に既に完了している）。
//
// contains（構成関係）には同じ含意が無いため対象外。全体をコンテナに含めても中身が
// 完了しているとは限らない、というflat DAG原則（進捗リング実装時の判断）と対称的な設計。
func CascadeSatisfyRequires(g *model.Graph, nodeID string, seen map[string]bool) {
	if seen[nodeID] {
		return
	}
	seen[nodeID] = true

	node, ok := g.Nodes[nodeID]
	if !ok {
		return
	}
	for _, reqID := range node.Requires {
		if reqNode, ok := g.Nodes[reqID]; ok {
			reqNode.Satisfied = true
			CascadeSatisfyRequires(g, reqID, seen)
		}
	}
}

// CascadeUnsatisfyDependents は CascadeSatisfyRequires と逆向きの走査。
// あるノードが未達に戻った時、そのノードを requires で指定している依存先（後工程）ノードも
// 再帰的に未達へ巻き戻す。「前提が崩れたなら、それに依存していた後工程の完了実績も
// 本当は成立していない」という requires の意味論に基づく
// （例: Natah を取り消すなら、Natahを前提にしていた Steel Path 解放も未達に戻る）。
//
// 前提側（CascadeSatisfyRequiresの対象）は触らない — Natahを取り消してもThe War Within等の
// 完了実績自体は消えないため。
func CascadeUnsatisfyDependents(g *model.Graph, nodeID string, seen map[string]bool) {
	if seen[nodeID] {
		return
	}
	seen[nodeID] = true

	for id, node := range g.Nodes {
		for _, reqID := range node.Requires {
			if reqID == nodeID {
				node.Satisfied = false
				CascadeUnsatisfyDependents(g, id, seen)
				break
			}
		}
	}
}

// CollectMembers は指定ノードから contains と requires の両方を再帰的に辿り、
// 関連する全ノードIDを集める（対象ノード自身は含まない）。
//
// contains のみを辿ると、真の Next Action（例: シンジケートランク上げ）が
// レポートから欠落する（Obsidian Dashboard.md 実装時に発見したギャップ）ため、
// 両エッジを統合して走査する。
func CollectMembers(g *model.Graph, nodeID string, seen map[string]bool) []string {
	if seen[nodeID] {
		return nil
	}
	seen[nodeID] = true

	node, ok := g.Nodes[nodeID]
	if !ok {
		return nil
	}

	members := []string{nodeID}
	for _, childID := range append(append([]string{}, node.Contains...), node.Requires...) {
		members = append(members, CollectMembers(g, childID, seen)...)
	}
	return members
}

// NodeView はノード本体（requires/contains含む）に、算出済みStateを添えたもの。
// フロントエンドがグラフ描画（レイアウト計算・エッジ描画）に必要な情報を、
// エッジ走査ロジックを再実装せずに一度で取得できるようにする。
type NodeView struct {
	*model.Node
	State model.NodeState `json:"state"`
}

type NextActionReport struct {
	BuildID    string              `json:"buildId"`
	Progress   Progress            `json:"progress"`
	Actionable []string            `json:"actionable"`
	Blocked    []string            `json:"blocked"`
	Satisfied  []string            `json:"satisfied"`
	Nodes      map[string]NodeView `json:"nodes"` // buildID自身を含む、メンバー全ノードのビュー
}

type Progress struct {
	Done  int `json:"done"`
	Total int `json:"total"`
}

// DeriveNextActions は buildID を起点に、状態別（ACTIONABLE/BLOCKED/SATISFIED）に
// 関連ノードを分類したレポートを返す。
func DeriveNextActions(g *model.Graph, buildID string) NextActionReport {
	seen := make(map[string]bool)
	members := CollectMembers(g, buildID, seen)

	report := NextActionReport{BuildID: buildID, Nodes: make(map[string]NodeView, len(members)+1)}

	if buildNode, ok := g.Nodes[buildID]; ok {
		report.Nodes[buildID] = NodeView{Node: buildNode, State: model.StateSatisfied}
	}

	for _, id := range members {
		if id == buildID {
			continue
		}
		node := g.Nodes[id]
		state := ResolveState(g, id, map[string]bool{})
		report.Nodes[id] = NodeView{Node: node, State: state}

		switch state {
		case model.StateSatisfied:
			report.Satisfied = append(report.Satisfied, id)
		case model.StateActionable:
			report.Actionable = append(report.Actionable, id)
		default:
			report.Blocked = append(report.Blocked, id)
		}
	}
	report.Progress = Progress{
		Done:  len(report.Satisfied),
		Total: len(report.Satisfied) + len(report.Actionable) + len(report.Blocked),
	}
	return report
}

File: ts/server/engine.ts
// Port of pkg/engine/engine.go — the cascade/resolve semantics that are the
// conceptual core of the app. CODEOWNERS-protects pkg/engine + pkg/model in
// the Go repo (required review, no direct-to-main push); once this file
// exists it should get the same protection (see migration plan guardrails —
// noted here as a to-do, not applied by this port itself).

import type { Graph, Node, NodeState } from "./model.ts";

/**
 * Recursively resolves a node's fulfillment state. requires empty, or all
 * SATISFIED, → ACTIONABLE. A cycle is treated as BLOCKED rather than
 * recursing forever.
 */
export function resolveState(
  g: Graph,
  nodeId: string,
  stack: ReadonlySet<string> = new Set(),
): NodeState {
  const node = g.nodes[nodeId];
  if (!node) return "BLOCKED";
  if (node.satisfied) return "SATISFIED";
  if (stack.has(nodeId)) return "BLOCKED"; // cycle guard
  if (node.requires.length === 0) return "ACTIONABLE";

  const next = new Set(stack);
  next.add(nodeId);

  for (const reqId of node.requires) {
    if (resolveState(g, reqId, next) !== "SATISFIED") return "BLOCKED";
  }
  return "ACTIONABLE";
}

/**
 * When a node becomes satisfied, walks its `requires` chain backward and
 * marks everything in it satisfied too — "if the later step is done, the
 * earlier steps it required must already be done" (e.g. completing Natah
 * implies The War Within and Saya's Vigil are already done).
 *
 * `contains` carries no such implication (a build being satisfied doesn't
 * mean its parts are — that's the flat-DAG design behind the progress
 * ring), so it is not walked here.
 */
export function cascadeSatisfyRequires(g: Graph, nodeId: string, seen: Set<string> = new Set()): void {
  if (seen.has(nodeId)) return;
  seen.add(nodeId);

  const node = g.nodes[nodeId];
  if (!node) return;
  for (const reqId of node.requires) {
    const reqNode = g.nodes[reqId];
    if (reqNode) {
      reqNode.satisfied = true;
      cascadeSatisfyRequires(g, reqId, seen);
    }
  }
}

/**
 * The reverse walk of {@link cascadeSatisfyRequires}. When a node reverts to
 * unsatisfied, every node that lists it in `requires` (its downstream
 * dependents) reverts too, recursively — "if the prerequisite no longer
 * holds, anything built on top of it wasn't really done either" (e.g.
 * un-completing Natah reverts Steel Path access that required it).
 *
 * The prerequisite side is left untouched — un-completing Natah doesn't
 * erase the fact The War Within was finished.
 */
export function cascadeUnsatisfyDependents(g: Graph, nodeId: string, seen: Set<string> = new Set()): void {
  if (seen.has(nodeId)) return;
  seen.add(nodeId);

  for (const [id, node] of Object.entries(g.nodes)) {
    if (node.requires.includes(nodeId)) {
      node.satisfied = false;
      cascadeUnsatisfyDependents(g, id, seen);
    }
  }
}

/**
 * Recursively walks both `contains` and `requires` from nodeId and collects
 * every related node id (not including nodeId itself). Walking `contains`
 * alone drops real Next Actions (e.g. a syndicate-rank requirement) from the
 * report — a gap found while building the original Obsidian prototype — so
 * both edge kinds are combined here.
 */
export function collectMembers(g: Graph, nodeId: string, seen: Set<string> = new Set()): string[] {
  if (seen.has(nodeId)) return [];
  seen.add(nodeId);

  const node = g.nodes[nodeId];
  if (!node) return [];

  const members = [nodeId];
  for (const childId of [...node.contains, ...node.requires]) {
    members.push(...collectMembers(g, childId, seen));
  }
  return members;
}

/** A node's own fields plus its computed state, for the frontend to draw
 * layout/edges from one fetch without reimplementing traversal. */
export interface NodeView extends Node {
  state: NodeState;
}

export interface Progress {
  done: number;
  total: number;
}

export interface NextActionReport {
  buildId: string;
  progress: Progress;
  actionable: string[];
  blocked: string[];
  satisfied: string[];
  /** Every member node's view, including buildId itself. */
  nodes: Record<string, NodeView>;
}

/** Classifies every node related to buildId (via collectMembers) into
 * ACTIONABLE/BLOCKED/SATISFIED and returns a report the frontend renders
 * from directly. */
export function deriveNextActions(g: Graph, buildId: string): NextActionReport {
  const members = collectMembers(g, buildId, new Set());

  const report: NextActionReport = {
    buildId,
    progress: { done: 0, total: 0 },
    actionable: [],
    blocked: [],
    satisfied: [],
    nodes: {},
  };

  const buildNode = g.nodes[buildId];
  if (buildNode) {
    // The root is always shown as satisfied ("起点"/origin), regardless of
    // its own `satisfied` field — matches the Go original exactly.
    report.nodes[buildId] = { ...buildNode, state: "SATISFIED" };
  }

  for (const id of members) {
    if (id === buildId) continue;
    const node = g.nodes[id];
    // collectMembers only ever adds ids that resolved to a real node, so
    // this is unreachable in practice — kept only to satisfy
    // noUncheckedIndexedAccess.
    if (!node) continue;
    const state = resolveState(g, id, new Set());
    report.nodes[id] = { ...node, state };

    switch (state) {
      case "SATISFIED":
        report.satisfied.push(id);
        break;
      case "ACTIONABLE":
        report.actionable.push(id);
        break;
      default:
        report.blocked.push(id);
        break;
    }
  }

  report.progress = {
    done: report.satisfied.length,
    total: report.satisfied.length + report.actionable.length + report.blocked.length,
  };
  return report;
}
